### PR TITLE
Refactor extract-partitions

### PR DESCRIPTION
**Pull request by bocajnotnef**
Generators galore! Classes are fun! Wooo!

Ref #144 (man, that's old)

TO-DO:
 - [x] Pull out all actual processing into classes/generators and just let the script handle IO, etc.
 - [ ] Waffle: Move stuff into oxli?
 - [x] Shove things like min_size in init when we build the extractor

### CODE REVIEW

**Comment by bocajnotnef** [comment]
- [ ] Is it mergeable?
- [ ] Did it pass the tests?
- [ ] If it introduces new functionality in scripts/ is it tested?
  Check for code coverage with `make clean diff-cover`
- [ ] Is it well formatted? Look at `make pep8`, `make diff_pylint_report`,
  `make cppcheck`, and `make doc` output. Use `make format` and manual
  fixing as needed.
- [ ] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [ ] Is it documented in the ChangeLog?
  http://en.wikipedia.org/wiki/Changelog#Format
- [ ] Was a spellchecker run on the source code and documentation after
  changes were made?
- [ ] Is the Copyright year up to date?

**Comment by bocajnotnef** [comment]
@ctb @camillescott @luizirber CR and thoughts on migrating the generators to oxli?

**Comment by luizirber** [comment]
Check https://github.com/dib-lab/khmer/compare/rfact/extract-part...rfact/extract-part_sugg for a suggestion on how to define the generators and context managers. Defining the `__call__` of a class works, but it looks a bit weird to me...

File: scripts/extract-partitions.py [hunk 1]
```diff
@@ -83,15 +84,159 @@ def get_parser():
     return parser
 
 
-# pylint: disable=too-many-statements
-def main():  # pylint: disable=too-many-locals,too-many-branches
+class PartitonedReadIterator(object):
```
Comment (ctb) [inline scripts/extract-partitions.py]: spelling, please ;)

"Partitioned"
"Generator"

File: tests/test_scripts.py [hunk 2]
```diff
@@ -1349,6 +1349,34 @@ def test_extract_partitions():
     assert len(parts) == 1, len(parts)
 
 
+def test_extract_paired_inconsitent_formats():
```
Comment (ctb) [inline tests/test_scripts.py]: "inconsistent"

**Comment by bocajnotnef** [comment]
@luizirber The tricky thing is in some cases I need to do the setup in __init__, which I don't think is possible when using the decorator.

Also, cleaned up, tested, ready for merge (unless we want to migrate stuff to oxli)

**Comment by luizirber** [comment]
@bocajnotnef you can do setup before the yield and teardown after the yield, what kind of things do you need to save for later?

**Comment by bocajnotnef** [comment]
We pull statistics from the ReadGenerator, here; https://github.com/dib-lab/khmer/pull/1251/files#diff-2b097bf3eed5cf5518cea07f5c4fb00eR316

**Comment by luizirber** [comment]
I see. But can you use `__iter__` instead of `__call__`? Similar to what we do in [screed][1], for example.

[1]: https://github.com/dib-lab/screed/blob/ddcbcbd7618efb3a90064fae4531140387e62b52/screed/openscreed.py#L115

**Comment by bocajnotnef** [comment]
I don't feel like that'd clean it up any, and as it is the functionality seems clear. (Plus, we did it this way in normalize-by-median_

**Comment by ctb** [comment]
What's the reason to migrate stuff to oxli?

**Comment by bocajnotnef** [comment]
Just seemed logical since what's essentially happening is the script is
being broken down into functions that do the processing and a main function
that makes it run as a script.

**Comment by standage** [comment]
test this please jenkins

**Comment by standage** [comment]
LGTM.

**Comment by standage** [comment]
Thanks @bocajnotnef!

**Pull request merged**